nixVersions.stable: nix_2_28 -> nix_2_31

philiptaron opened the pull request:
NixCon 2025 has been a wonderful experience. Let's make it even better for the whole world by giving them a new, fresh, well-tested Nix version. 

Per [the readme](https://github.com/NixOS/nixpkgs/blob/a4716b0e8b18ffb120aff902016f6e846e50bd2f/pkgs/tools/package-management/nix/README.md#L31-L35), we're targeting staging.

Previous bump:
- https://github.com/NixOS/nixpkgs/pull/375820

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

philiptaron committed:
nixVersions.stable: nix_2_28 -> nix_2_31

philiptaron committed:
nix-serve: update to componentized Nix

Co-authored-by: Guillame Maudoux <layus.on@gmail.com>

philiptaron commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 440681 --checkout commit`
Commit: `62f82d6a8f43914978b6dd310d6f060183ab24b7`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package marked as broken and skipped:</summary>
  <ul>
    <li>tests.pkg-config.defaultPkgConfigPackages.tests-combined</li>
  </ul>
</details>
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 130 packages built:</summary>
  <ul>
    <li>appvm</li>
    <li>bower2nix</li>
    <li>bundix</li>
    <li>cabal2nix</li>
    <li>cached-nix-shell</li>
    <li>chirpstack-concentratord</li>
    <li>colmena</li>
    <li>comma</li>
    <li>common-updater-scripts</li>
    <li>crate2nix</li>
    <li>crush</li>
    <li>disko</li>
    <li>dub-to-nix</li>
    <li>dydisnix</li>
    <li>fusionInventory</li>
    <li>gcalcli</li>
    <li>gcalcli.dist</li>
    <li>gclient2nix</li>
    <li>glpi-agent</li>
    <li>haskellPackages.cabal2nix-unstable</li>
    <li>haskellPackages.cli-nix</li>
    <li>haskellPackages.cli-nix.doc</li>
    <li>haskellPackages.niv</li>
    <li>haskellPackages.niv.bin</li>
    <li>haskellPackages.niv.data</li>
    <li>haskellPackages.niv.doc</li>
    <li>haskellPackages.nix-paths</li>
    <li>haskellPackages.nix-paths.doc</li>
    <li>haskellPackages.nix-thunk</li>
    <li>haskellPackages.nix-thunk.doc</li>
    <li>haskellPackages.nvfetcher</li>
    <li>haskellPackages.nvfetcher.doc</li>
    <li>haskellPackages.update-nix-fetchgit</li>
    <li>haskellPackages.update-nix-fetchgit.doc</li>
    <li>home-manager</li>
    <li>kcl</li>
    <li>libnixxml</li>
    <li>lua51Packages.luarocks-nix</li>
    <li>luarocks-nix (luaPackages.luarocks-nix)</li>
    <li>lua53Packages.luarocks-nix</li>
    <li>lua54Packages.luarocks-nix</li>
    <li>luajitPackages.luarocks-nix</li>
    <li>luarocks-packages-updater</li>
    <li>luarocks-packages-updater.dist</li>
    <li>lux-cli</li>
    <li>mdwatch</li>
    <li>mlv-app</li>
    <li>nil</li>
    <li>nim_lk</li>
    <li>niv (niv.bin, niv.data)</li>
    <li>nix-bundle</li>
    <li>nix-direnv</li>
    <li>nix-forecast</li>
    <li>nix-index</li>
    <li>nix-init</li>
    <li>nix-pin</li>
    <li>nix-prefetch</li>
    <li>nix-prefetch-docker</li>
    <li>nix-prefetch-github (python313Packages.nix-prefetch-github)</li>
    <li>nix-prefetch-github.dist (python313Packages.nix-prefetch-github.dist)</li>
    <li>nix-prefetch-github.man (python313Packages.nix-prefetch-github.man)</li>
    <li>nix-required-mounts</li>
    <li>nix-required-mounts.dist</li>
    <li>nix-serve</li>
    <li>nix-template</li>
    <li>nix-update</li>
    <li>nix-update.dist</li>
    <li>nix-visualize</li>
    <li>nix-visualize.dist</li>
    <li>nix-web</li>
    <li>nixStatic</li>
    <li>nixStatic.dev</li>
    <li>nixStatic.doc</li>
    <li>nixStatic.man</li>
    <li>nixci</li>
    <li>nixos-anywhere</li>
    <li>nixos-generators</li>
    <li>nixos-option</li>
    <li>nixos-rebuild</li>
    <li>nixos-rebuild-ng</li>
    <li>nixos-rebuild-ng.dist</li>
    <li>nixos-shell</li>
    <li>nixpkgs-hammering</li>
    <li>nixpkgs-manual</li>
    <li>nixpkgs-review</li>
    <li>nixpkgs-review.dist</li>
    <li>nixpkgs-reviewFull</li>
    <li>nixpkgs-reviewFull.dist</li>
    <li>nixseparatedebuginfod2</li>
    <li>nixtract</li>
    <li>node2nix (nodePackages.node2nix)</li>
    <li>nodePackages_latest.node2nix</li>
    <li>nps</li>
    <li>nuget-to-json</li>
    <li>nurl</li>
    <li>nvfetcher</li>
    <li>python312Packages.nix-kernel</li>
    <li>python312Packages.nix-kernel.dist</li>
    <li>python312Packages.nix-prefetch-github</li>
    <li>python312Packages.nix-prefetch-github.dist</li>
    <li>python312Packages.nix-prefetch-github.man</li>
    <li>python313Packages.nix-kernel</li>
    <li>python313Packages.nix-kernel.dist</li>
    <li>python313Packages.nixpkgs-updaters-library</li>
    <li>python313Packages.nixpkgs-updaters-library.dist</li>
    <li>ragenix</li>
    <li>sbomnix</li>
    <li>sbomnix.dist</li>
    <li>sus-compiler</li>
    <li>swiftpm2nix (swiftPackages.swiftpm2nix)</li>
    <li>terranix</li>
    <li>tests.devShellTools.nixos</li>
    <li>tests.pkg-config.defaultPkgConfigPackages.nix-cmd</li>
    <li>tests.pkg-config.defaultPkgConfigPackages.nix-expr</li>
    <li>tests.pkg-config.defaultPkgConfigPackages.nix-main</li>
    <li>tests.pkg-config.defaultPkgConfigPackages.nix-store</li>
    <li>tests.testers.lycheeLinkCheck.network</li>
    <li>tests.testers.nixosTest-example</li>
    <li>tests.testers.runNixOSTest-example (tests.testers.runNixOSTest-extendNixOS)</li>
    <li>tests.trivial-builders.references</li>
    <li>update-nix-fetchgit</li>
    <li>update-python-libraries</li>
    <li>vimPluginsUpdater</li>
    <li>vulnix</li>
    <li>vulnix.dist</li>
    <li>vulnix.doc</li>
    <li>vulnix.man</li>
    <li>wp4nix</li>
    <li>xosd-xft</li>
    <li>zon2nix</li>
  </ul>
</details>

philiptaron commented:
I'd appreciate any testing on macOS.

MattSturgeon reviewed (approved):
Not tested on my end, but no objections either :+1:

Approved pending additional reviews with testing and/or more expertise.

philiptaron commented:
> is there a perticualr reason to target staging? the rebuild count doesn't seem that high. and i can't find any history of `nixVersions.stable` changes targeting staging. (also can't find the readme in descrption)

I should have linked it.

https://github.com/NixOS/nixpkgs/blob/a4716b0e8b18ffb120aff902016f6e846e50bd2f/pkgs/tools/package-management/nix/README.md#L31-L35

Evidently it has to do with [rebuilds of the NixOS test driver](https://github.com/NixOS/nixpkgs/issues/437208).

philiptaron commented:
I upgraded my macOS system to Nix 2.31.1 to produce the above build results. Nothing looks out of place. `nixStatic` should be marked broken on macOS.

philiptaron commented:
Nix 2.31.2 has been merged with the fixes to the above.

roberth commented:
IIUC this needs a rebase and a fallback paths update.

vcunat reviewed (commented):
Ther rebuilds certainly won't be an issue.

lovesegfault commented:
Approved, but deferring the merge to @vcunat

Mic92 commented:
```
% nix-build --eval-system aarch64-darwin -A nixVersions.stable.tests
Finished at 19:35:16 after 0s
/nix/store/sy8vn765njh5rkn4z3pf3k3hndrzahsl-test-nix-fallback-paths-version-equals-nix-stable
/nix/store/9zh2wrpyq2cipry8wqqw74lraam76g0k-nixpkgs-lib-tests-nix-2.31.2
/nix/store/h94y4h0nwh60ylx5psypa43wlaghgjbk-check-pkg-config-nix
/nix/store/5iwsfrlbyfirmmfn7l1wir61lq4h00m5-nix-src-version
```

Mic92 committed:
remove nixStatic from nix.passthru.tests

Mic92 commented:
removed nixStatic from the tests, I might also drop nixStatic from nixpkgs because the pkgsStatic maintainance is not good enough and it always seems I have to fix all kinds of dependencies for a new nix release each time.

misuzu commented:
The cross build is broken, at least for riscv64:
```
% nix-build . -A pkgsCross.riscv64.nixVersions.stable
these 15 derivations will be built:
  /nix/store/bqsw97jh96grs29l689wxcn1gppx1wax-nix-util-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/48jkgwbsynx67mdckwa7gvm7sj43m66i-nix-util-c-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/qjn9az0xwm2h3gd6mxvan6057aqdf35y-nix-store-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/259a5lfdnzj99l09zq9gz2lk3blkv6m5-nix-store-c-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/vkvpjzsgn6fm9khr5kc1bkl727gjdg7h-nix-fetchers-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/qdvwzp6xl2c8kdj48sk3xdvi2q56dc22-nix-expr-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/c8whlmmh7lvgs52w6ry40vfagd0vhmsb-nix-flake-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/nkphrhy9g0ybhg8iq8adnrbbwdms43j8-nix-main-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/3kdzab4q6rmfj452fpx6i2bm7caj4yfv-nix-cmd-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/x3s1a7nrvhfw9941y3s22a8s4gkf42h4-nix-expr-c-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/r3qkmgvk90bs9slp6nrfaay1zx93bvb1-nix-fetchers-c-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/g2cdhz0afxiw9z5l094g08ik4w5gcmd2-nix-flake-c-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/hrhsgb0xwz71qham6lk86pa6jnigzcyn-nix-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/psjlcml63bsy8qk274fwm5fh0vhqjjhl-nix-main-c-riscv64-unknown-linux-gnu-2.31.2.drv
  /nix/store/d3bh3fcrsyg7qsdpi6fl01hg8fj8nyki-nix-riscv64-unknown-linux-gnu-2.31.2.drv
building '/nix/store/bqsw97jh96grs29l689wxcn1gppx1wax-nix-util-riscv64-unknown-linux-gnu-2.31.2.drv'...
Running phase: unpackPhase
unpacking source archive /nix/store/z4s453vx7vndag81slkdxxw144rk5l0m-source
source root is source/./src/libutil
separate-debug-info: recording checksum of source files for debug support...
Running phase: patchPhase
Running phase: updateAutotoolsGnuConfigScriptsPhase
Running phase: updateAutotoolsGnuConfigScriptsPhase
Running phase: configurePhase
mesonConfigurePhase flags: --prefix=/nix/store/jzjadrwyyd5xixlh5788n6w761ww30pl-nix-util-riscv64-unknown-linux-gnu-2.31.2 --libdir=/nix/store/jzjadrwyyd5xixlh5788n6w761ww30pl-nix-util-riscv64-unknown-linux-gnu-2.31.2/lib --libexecdir=/nix/store/jzjadrwyyd5xixlh5788n6w761ww30pl-nix-util-riscv64-unknown-linux-gnu-2.31.2/libexec --bindir=/nix/store/jzjadrwyyd5xixlh5788n6w761ww30pl-nix-util-riscv64-unknown-linux-gnu-2.31.2/bin --sbindir=/nix/store/jzjadrwyyd5xixlh5788n6w761ww30pl-nix-util-riscv64-unknown-linux-gnu-2.31.2/sbin --includedir=/nix/store/zbkhwkij6xpz3nhdpk8mlwh4fa81zqd1-nix-util-riscv64-unknown-linux-gnu-2.31.2-dev/include --mandir=/nix/store/jzjadrwyyd5xixlh5788n6w761ww30pl-nix-util-riscv64-unknown-linux-gnu-2.31.2/share/man --infodir=/nix/store/jzjadrwyyd5xixlh5788n6w761ww30pl-nix-util-riscv64-unknown-linux-gnu-2.31.2/share/info --localedir=/nix/store/jzjadrwyyd5xixlh5788n6w761ww30pl-nix-util-riscv64-unknown-linux-gnu-2.31.2/share/locale -Dauto_features=enabled -Dwrap_mode=nodownload --buildtype=release --cross-file=/nix/store/ixrssm87rrigpjn42awcvf37z79viq73-cross-file.conf -Dcpuid=disabled -Db_lto=true
The Meson build system
Version: 1.9.0
Source dir: /build/source/src/libutil
Build dir: /build/source/src/libutil/build
Build type: cross build
Project name: nix-util
Project version: 2.31.2

meson.build:1:0: ERROR: Unable to detect linker for compiler `riscv64-unknown-linux-gnu-g++ -Wl,--version -fuse-ld=gold`
stdout:
stderr: collect2 version 14.3.0
[cannot find ld] -plugin /nix/store/78w7scn9l5rcsphwmkwysv1vyzgr765z-riscv64-unknown-linux-gnu-gcc-14.3.0/libexec/gcc/riscv64-unknown-linux-gnu/14.3.0/liblto_plugin.so -plugin-opt=/nix/store/78w7scn9l5rcsphwmkwysv1vyzgr765z-riscv64-unknown-linux-gnu-gcc-14.3.0/libexec/gcc/riscv64-unknown-linux-gnu/14.3.0/lto-wrapper -plugin-opt=-fresolution=/build/ccnyvVAo.res -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc --eh-frame-hdr -melf64lriscv -X -dynamic-linker /lib/ld-linux-riscv64-lp64d.so.1 -pie /nix/store/8rxkb9z81ypvmaidhdm8d15fjy3nmi28-glibc-riscv64-unknown-linux-gnu-2.40-66/lib/Scrt1.o /nix/store/8rxkb9z81ypvmaidhdm8d15fjy3nmi28-glibc-riscv64-unknown-linux-gnu-2.40-66/lib/crti.o /nix/store/78w7scn9l5rcsphwmkwysv1vyzgr765z-riscv64-unknown-linux-gnu-gcc-14.3.0/lib/gcc/riscv64-unknown-linux-gnu/14.3.0/crtbeginS.o -L/nix/store/8rxkb9z81ypvmaidhdm8d15fjy3nmi28-glibc-riscv64-unknown-linux-gnu-2.40-66/lib -L/nix/store/m616w6krvnq9d8wnpnkhr7w7g9nzsn46-riscv64-unknown-linux-gnu-gcc-14.3.0-lib/riscv64-unknown-linux-gnu/lib -L/nix/store/78w7scn9l5rcsphwmkwysv1vyzgr765z-riscv64-unknown-linux-gnu-gcc-14.3.0/lib/gcc/riscv64-unknown-linux-gnu/14.3.0 -L/nix/store/78w7scn9l5rcsphwmkwysv1vyzgr765z-riscv64-unknown-linux-gnu-gcc-14.3.0/lib/gcc/riscv64-unknown-linux-gnu/14.3.0/../../../../riscv64-unknown-linux-gnu/lib --version -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc /nix/store/78w7scn9l5rcsphwmkwysv1vyzgr765z-riscv64-unknown-linux-gnu-gcc-14.3.0/lib/gcc/riscv64-unknown-linux-gnu/14.3.0/crtendS.o /nix/store/8rxkb9z81ypvmaidhdm8d15fjy3nmi28-glibc-riscv64-unknown-linux-gnu-2.40-66/lib/crtn.o
collect2: fatal error: cannot find ‘ld’
compilation terminated.


A full log can be found at /build/source/src/libutil/build/meson-logs/meson-log.txt
```

Mic92 commented:
```
% nix-build --eval-system aarch64-linux -A nixVersions.stable.tests --keep-going
Finished at 21:33:53 after 21s
/nix/store/ndmcrr5h3m91rkhb1c3mrg5k5hxrqgnf-vm-test-run-nix-misc
/nix/store/yg5gkhnfrskc5cs37nkgnr0waq9s425p-test-nix-fallback-paths-version-equals-nix-stable
/nix/store/w5ciz0qrn28phr6r1ws28w2in02y9lbn-nixpkgs-lib-tests-nix-2.31.2
/nix/store/bdn6qsbrmy32gz07arydicma2rq2lsvb-check-pkg-config-nix
/nix/store/l7sysgkn2j0h18fly3v5fg84cj1jzq7d-vm-test-run-installer-simpleUefiSystemdBoot
/nix/store/fvr0r16yyaaszs60c4kc66gr4wj1p7n4-nix-src-version
/nix/store/22zzzhkh9r98r5bvvfk5rzv0c5malkcs-vm-test-run-nix-upgrade-2.31.2-2.31.2
nom-build --eval-system aarch64-linux -A nixVersions.stable.tests --keep-goin  8.72s user 2.27s system 51% cpu 21.510 total
```

Mic92 committed:
nix: remove deprecated fuse-ld=gold linker

Binutils deprecated the gold linker, so do we.
We already did this a while ago in the nix repository since we switched to mold for development.

misuzu reviewed (approved):
The cross build to riscv64 is fixed, thanks!
```
% nix-build . -A pkgsCross.riscv64.nixVersions.stable
/nix/store/qazd6knqxv2a0d5gz76m56y5gs3yzcq4-nix-riscv64-unknown-linux-gnu-2.31.2
```

philiptaron commented:
Hell yeah. This is going to be a staging-next for the ages! And if the 2.32 release (with the awesome memory wins) happens soon enough, we might be able to sneak that into 25.11 as the stable Nix.

Izorkin commented:
@xokdvium not working this variant:
```
  nixpkgs.overlays = [
    nix = pkgs.nixVersions.nix_2_31.overrideScope (final: prev: { nix-store = prev.nix-store.override { withAWS = false; }; })
  ];
```

Or am I doing something wrong.

xokdvium commented:
Presumably something like this should work. There's 2 layers of overrides here, one is the overlay and the other is the overrideScope.

```nix
nixpkgs.overlays = [
  (_: prevPkgs: { nix = prevPkgs.nixVersions.stable.overrideScope (final: prev: { nix-store = prev.nix-store.override { withAWS = false; }; } }))
];
```

xokdvium commented:
Sorry, I was typing on the phone. Just fix the parenthesis, but the gist is that. You can just look up how to specify overlays in the relevant nixpkgs documentation.
